dm/master/scheduler/scheduler.go

```diff
@@ -1777,3 +1934,208 @@ func (s *Scheduler) SetWorkerClientForTest(name string, mockCli workerrpc.Client
+					}
+				}
+				retryNum++
+			}
+		} else {
+			if err != nil {
+				log.L().Error("observeLoadTask is failed and will quit now", zap.Error(err))
+			} else {
+				log.L().Info("observeLoadTask will quit now")
+			}
+			return err
+		}
+	}
+}
+
+// RemoveLoadTask removes the loadtask by task.
+func (s *Scheduler) RemoveLoadTask(task string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+	_, _, err := ha.DelLoadTaskByTask(s.etcdCli, task)
+	if err != nil {
+		return err
+	}
+	delete(s.loadTasks, task)
+	return nil
+}
+
+// getTransferWorkerAndSource tries to get transfer worker and source.
+// worker, source	This means a subtask finish load stage, often called by handleLoadTaskDel.
+// worker, ""		This means a free worker online, often called by tryBoundForWorker.
+// "", source		This means a unbounded source online, often called by tryBoundForSource.
+func (s *Scheduler) getTransferWorkerAndSource(worker, source string) (string, string) {
+	// origin worker not free, try to get a source.
+	if worker != "" {
+		// try to get a unbounded source
+		for sourceID := range s.unbounds {
+			if sourceID != source && s.hasLoadTaskByWorkerAndSource(worker, sourceID) {
+				return "", sourceID
+			}
+		}
+		// try to get a bounded source
+		for sourceID, w := range s.bounds {
+			if sourceID != source && s.hasLoadTaskByWorkerAndSource(worker, sourceID) && !s.hasLoadTaskByWorkerAndSource(w.baseInfo.Name, sourceID) {
+				return w.baseInfo.Name, sourceID
+			}
+		}
+	}
+
+	// origin source bounded, try to get a worker
+	if source != "" {
+		// try to get a free worker
+		for _, w := range s.workers {
+			workerName := w.baseInfo.Name
+			if workerName != worker && w.Stage() == WorkerFree && s.hasLoadTaskByWorkerAndSource(workerName, source) {
+				return workerName, ""
+			}
+		}
+
+		// try to get a bounded worker
+		for _, w := range s.workers {
+			workerName := w.baseInfo.Name
+			if workerName != worker && w.Stage() == WorkerBound {
+				if s.hasLoadTaskByWorkerAndSource(workerName, source) && !s.hasLoadTaskByWorkerAndSource(workerName, w.bound.Source) {
+					return workerName, w.bound.Source
+				}
+			}
+		}
+	}
+
+	return "", ""
+}
+
+// hasLoadTaskByWorkerAndSource check whether there is a load subtask for the worker and source.
+func (s *Scheduler) hasLoadTaskByWorkerAndSource(worker, source string) bool {
+	for _, sourceWorkerMap := range s.loadTasks {
+		if workerName, ok := sourceWorkerMap[source]; ok && workerName == worker {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *Scheduler) handleLoadTaskDel(loadTask ha.LoadTask) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.loadTasks[loadTask.Task]; !ok {
+		return nil
+	}
+	if _, ok := s.loadTasks[loadTask.Task][loadTask.Source]; !ok {
+		return nil
+	}
+
+	loadTask.Worker = s.loadTasks[loadTask.Task][loadTask.Source]
```

Comment:
this makes me confused because it modifies input, but input is not reference typed so no effect indeed 😵 
slightly recommend using a local variable to store worker and source of `loadTask`